test/shared/src/main/scala/zio/test/mock/MockClock.scala

```diff
@@ -50,34 +53,98 @@ object MockClock {
       clockState.get.map(_.nanoTime)
 
     final def sleep(duration: Duration): UIO[Unit] =
-      adjust(duration) *> clockState.update(data => data.copy(sleeps0 = duration :: data.sleeps0)).unit
-
-    val sleeps: UIO[List[Duration]] = clockState.get.map(_.sleeps0.reverse)
+      for {
+        latch <- Promise.make[Nothing, Unit]
+        await <- clockState.modify { data =>
+                  if (duration > Duration.fromNanos(data.nanoTime))
+                    (true, data.copy(sleeps = (duration, latch) :: data.sleeps))
+                  else
+                    (false, data)
+                }
+        _ <- if (await) latch.await else latch.succeed(())
+      } yield ()
+
+    val sleeps: UIO[List[Duration]] = clockState.get.map(_.sleeps.map(_._1))
 
     final def adjust(duration: Duration): UIO[Unit] =
       clockState.update { data =>
         Data(
           data.nanoTime + duration.toNanos,
           data.currentTimeMillis + duration.toMillis,
-          data.sleeps0,
+          data.sleeps,
           data.timeZone
         )
-      }.unit
+      } *> wakeUp
 
     final def setTime(duration: Duration): UIO[Unit] =
-      clockState.update(_.copy(nanoTime = duration.toNanos, currentTimeMillis = duration.toMillis)).unit
+      clockState.update { data =>
+        data.copy(
+          nanoTime = duration.toNanos,
+          currentTimeMillis = duration.toMillis
+        )
+      } *> wakeUp
 
     final def setTimeZone(zone: ZoneId): UIO[Unit] =
       clockState.update(_.copy(timeZone = zone)).unit
 
     val timeZone: UIO[ZoneId] =
       clockState.get.map(_.timeZone)
+
+    val scheduler: ZIO[Any, Nothing, IScheduler] =
+      ZIO.runtime[Any].flatMap { runtime =>
+        ZIO.succeed {
+          new IScheduler {
+            final def schedule(task: Runnable, duration: Duration): CancelToken =
+              duration match {
+                case Duration.Infinity =>
+                  ConstFalse
+                case Duration.Zero =>
+                  task.run()
+                  ConstFalse
+                case duration: Duration =>
+                  runtime.unsafeRun {
+                    for {
+                      latch <- Promise.make[Nothing, Unit]
+                      _     <- latch.await.flatMap(_ => ZIO.effect(task.run())).fork
+                      _ <- clockState.update { data =>
+                            data.copy(
+                              sleeps =
+                                (Duration.fromNanos(data.nanoTime) + duration, latch) ::
+                                  data.sleeps
+                            )
+                          }
+                    } yield () => runtime.unsafeRun(cancel(latch))
+                  }
+              }
+            final def shutdown(): Unit = ()
```

Comment:
Perhaps `shutdown` should run all pending tasks??? Not sure on that one.

Reply:
I wondered about that too. `java.util.concurrent.ScheduledExecutorService` does run pending tasks on shutdown so I updated to do the same.